voidcrew/code/modules/overmap/ships/pvp/subverter.dm

```diff
@@ -0,0 +1,235 @@
+	wires = new /datum/wires/subverter(src)
+	update_stats()
+	subverter_cooldown = world.time + sub_recharge
+
+/obj/machinery/subverter/connect_to_shuttle(obj/docking_port/mobile/port, obj/docking_port/stationary/dock)
+	ship = port.current_ship
+
+/obj/machinery/subverter/on_deconstruction()
+	//unlink autopilot
+	if (aux)
+		aux.sub = null
+		aux = null
+	..()
+
+/obj/machinery/subverter/proc/get_recharge_str()
+	. = (subverter_cooldown - world.time) / 10
+	. *= . > 0
+	return "[add_leading(num2text((. / 60) % 60), 2, "0")]:[add_leading(num2text(. % 60), 2, "0")]"
+
+/obj/machinery/subverter/proc/subvert_bark(message, soundin = SUBVERTER_BARK_FAILURE)
+	if (aux)
+		aux.say(message)
+	playsound(loc, soundin, 90, 1, 0)
+
+/*
+	This handles the different contexts a subverter could bark in response to.
+	There is also a single-line version of the logical operations in "can_subvert" (note that it excludes the antivirus check)
+
+	antag - Set to true if you want to ignore factions entirely
+*/
+/obj/machinery/subverter/proc/bark_processing(obj/structure/overmap/ship/simulated/target_ship, antag = FALSE, bark_success_message = "Agent was successful! Bringing [target_ship.name] out of hyperspace.")
+	. = FALSE
+	if (!antag)
+		if (target_ship.prefix == "NEU")
+			subvert_bark("Cannot subvert NEU vessels!")
+			return
+		if (ship)
+			if (ship.prefix == "NEU")
+				subvert_bark("NEU Vessels cannot use the subverter!")
+				return
+			if (ship.prefix != "KOS")
+				if (ship.prefix == target_ship.prefix)
+					subvert_bark("Cannot subvert allied vessels!")
+					return
+	if (world.time < subverter_cooldown)
+		subvert_bark("Subverter recharging!")
+		return
+	if (world.time < target_ship.engine_cooldown)
+		subvert_bark("Target vessel is already subverted.")
+		return
+	if (target_ship.state == OVERMAP_SHIP_IDLE)
+		subvert_bark("Target vessel is currently docked (outside of hyperspace).")
+		return
+	if (target_ship.state == OVERMAP_SHIP_DOCKING)
+		subvert_bark("Target vessel is currently docking.")
+		return
+	if (target_ship.state == OVERMAP_SHIP_UNDOCKING)
+		subvert_bark("Target vessel is currently undocking.")
+		return
+	if (world.time < target_ship.sub_grace)
+		subvert_bark("Cannot subvert that vessel currently (Grace period).")
+		return
+	if (target_ship.antivirus_nodes > 0)
+		subvert_bark("Agent failed! Reason: Antivirus")
+		return
+	if (!can_subvert(target_ship))
+		subvert_bark("Failed to subvert (Unknown error)")
+		return
+	subvert_bark(bark_success_message, SUBVERTER_BARK_SUCCESS)
+	return TRUE
+
+//quick maths
+/obj/machinery/subverter/proc/can_subvert(obj/structure/overmap/ship/simulated/target_ship, antag = FALSE)
+	return !(((!antag) && (target_ship.prefix == "NEU" || (ship && (ship.prefix == "NEU" || (ship.prefix != "KOS" && ship.prefix == target_ship.prefix))))) || world.time < subverter_cooldown || world.time < target_ship.engine_cooldown || world.time < target_ship.sub_grace || target_ship.state != OVERMAP_SHIP_FLYING)
```

Comment:
this is really gross, you should really try to avoid long boolean statements like this because they extremely hard to read.

Heres an idea:
use this proc to return an int for each different fail case, and in the above proc (`bark_processing`) switch through the returned int, and give the error message then, so that you don't have to check for the exact same errors in two different procs.
You should define all the ints of course, so that its understandable what errors are being sent back

Reply:
i was gonna say what if instead of a switch case, theres just a list that stores all the strings and the int indexes through it but that might be too much effort to also have that keep track of what sound to play